click/_compat.py

```diff
@@ -502,9 +502,33 @@ def open_stream(filename, mode='r', encoding=None, errors='strict',
     # as a proxy in the same folder and then using the fdopen
     # functionality to wrap it in a Python file.  Then we wrap it in an
     # atomic file that moves the file over on close.
-    import tempfile
-    fd, tmp_filename = tempfile.mkstemp(dir=os.path.dirname(filename),
-                                        prefix='.__atomic-write')
+    import errno
+    import random
+
+    try:
+        perm = os.stat(filename).st_mode
+    except OSError:
+        perm = None
+
+    flags = os.O_RDWR | os.O_CREAT | os.O_EXCL
+    if 'b' in mode:
+        flags |= getattr(os, 'O_BINARY', 0)
+
+    while True:
+        tmp_filename = os.path.join(
+            os.path.dirname(filename),
+            '.__atomic-write%08x' % (random.randrange(1 << 32),),
+        )
+        try:
+            fd = os.open(tmp_filename, flags, 0o666 if perm is None else perm)
+            break
+        except OSError as e:
+            if e.errno == errno.EEXIST:
```

Comment:
The errorhandling in the implementation of `mkstemp` seems to be more sophisticated, and I believe it would be good if it were ported over as well.

Reply:
See https://github.com/python/cpython/blob/c47c8ba2969c9faf1c036b3dc4b0a5cf0d3aa0cc/Lib/tempfile.py#L248

Reply:
Sure, done.